nixos/modules/services/x11/redshift.nix

```diff
@@ -88,51 +71,101 @@ in {
+      type = types.submodule {
+        freeformType = settingsFormat.type;
+      };
+
+      default = {};
+
+      example = {
+        redshift = {
+          temp-day = 5500;
+          temp-night = 3700;
+          brightness-day = 1.0;
+          brightness-night = 0.8;
+          fade = 1;
+        };
+        randr = {
+          screen = 0;
+        };
+      };
+
       description = ''
-        Additional command-line arguments to pass to
-        <command>redshift</command>.
+        The configuration to pass to Redshift/Gammastep.
+
+        Available options for Redshift described in
+        <citerefentry>
+          <refentrytitle>redshift</refentrytitle>
+          <manvolnum>1</manvolnum>
+        </citerefentry> and for Gammastep in
+        <citerefentry>
+          <refentrytitle>gammastep</refentrytitle>
+          <manvolnum>1</manvolnum>
+        </citerefentry>.
       '';
     };
   };
 
-  config = mkIf cfg.enable {
+  config = let
+    mainSection = if strings.hasInfix "gammastep" cfg.executable
+      # https://gitlab.com/chinstrap/gammastep/-/commit/1608ed61154cc652b087e85c4ce6125643e76e2f
+      then "general"
+      else "redshift";
+    target = if (cfg.settings.${mainSection}.adjustment-method or null) == "drm"
+      then "basic.target"
+      else "graphical-session.target";
+    configFile = settingsFormat.generate "redshift.conf" cfg.settings;
+    fullExecutable = "${cfg.package}${cfg.executable}";
+  in mkIf cfg.enable {
     # needed so that .desktop files are installed, which geoclue cares about
     environment.systemPackages = [ cfg.package ];
 
-    services.geoclue2.appConfig.redshift = {
+    services.geoclue2.appConfig.redshift = mkIf isGeoclue2 {
       isAllowed = true;
       isSystem = true;
     };
 
-    systemd.user.services.redshift =
-    let
-      providerString = if lcfg.provider == "manual"
-        then "${toString lcfg.latitude}:${toString lcfg.longitude}"
-        else lcfg.provider;
-    in
-    {
+    assertions = [
+      {
+        assertion = (cfg.settings ? "redshift") != (cfg.settings ? "general");
+        message = ''Both [redshift] and [general] sections found in settings. For gammastep use "general = {}" instead of "redshift = {}" in "server.redshift.settings".'';
+      }
+      {
+        assertion = (pathExists fullExecutable);
+        message = ''Executable not found. Are you using "/bin/redshift" for "pkgs.redshift" or "/bin/gammastep" for "pkgs.gammastep"?'';
+      }
+      {
+        assertion = (cfg.settings ? ${mainSection}.dawn-time) == (cfg.settings ? ${mainSection}.dusk-time);
+        message = "Time of dawn and time of dusk must be provided together.";
+      }
+    ];
+
+    services.redshift.settings = {
+      ${mainSection}.location-provider = lcfg.provider;
+      manual = mkIf isManual {
+        lat = mkIf options.location.latitude.isDefined lcfg.latitude;
+        lon = mkIf options.location.longitude.isDefined lcfg.longitude;
```

Comment:
Don't check for them being defined here, because we want an error if they're not defined.

Reply:
No we don't. It is perfectly possible to use provider = `manual` and set `dawn-time` and `dusk-time` instead (this is what I do.

Reply:
Can location provider be set to `null` :thinking: ? If yes,  I think I can remove those `mkIf`s for latitude/longitude.

Reply:
Hum... Maybe checking if not `dawn-time` and `dusk-time` is set too would be better.

Reply:
Ah so by selecting `manual` you have to either use `lat`/`lon` OR `dawn-time`/`dusk-time`? What happens if both are set?

Reply:
`dawn-time`/`dusk-time` has priority, I just think it is silly to make the user must set `latitude`/`longitude` when they want to use `dawn-time`/`dusk-time`.

Reply:
What I am thinking is a check like this:

```nix
manual = mkIf isManual && (cfg.settings ? ${mainSection}.dawn-time) && (cfg.settings ? ${mainSection}.dusk-time)
```

It is quite big, but I think would work. WDYT @Infinisil :thinking: ?

Reply:
Well, this doesn't work, infinite recursion.

Reply:
Ok, going to remove the `mkIf`s. Can't think of a better solution and setting `location = { latitude = 0; longitude = 0; }` does the trick.

Reply:
I'll take a look at this again later, but this isn't a great solution at the moment. There's a mismatch between redshift's idea of "manual" and the location provider's idea of "manual". I think the location provider module might need some change.

Reply:
I also don't have a problem of having this merge as is and after those changes in `location` is done I end up fixing this. Like, it is not ideal to have to set location to use time options but it works.

BTW, I think `location` should support a `location.provider = null`. This would work in this case because if provider isn't set we just use full manual mode.